File: frontend/src/objects/shared/response/ErrorResponse.ts
import type { ApiMessageParams } from '@/objects/shared/ApiMessageParams'
import { fromApiMessageParamsContract } from '@/objects/shared/ApiMessageParams'
import { readNullable, readRecord, readString } from '@/objects/shared/PageResponse'

export type ErrorResponse = {
  code: string | null
  message: string | null
  params: ApiMessageParams
}

export function fromErrorResponseContract(value: unknown, label: string): ErrorResponse {
  const response = readRecord(value, label)
  return {
    code: readNullable(response.code, `${label} code`, readString),
    message: readNullable(response.message, `${label} message`, readString),
    params: fromApiMessageParamsContract(response.params, `${label} params`),
  }
}

File: backend/src/main/scala/shared/objects/transport/ErrorResponse.scala
package shared.objects.transport

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import shared.objects.ApiMessageParams

/** 标准错误响应体，承载可本地化 code、兜底文本和消息参数。 */
final case class ErrorResponse(
  code: Option[String],
  message: Option[String],
  params: ApiMessageParams
)

/** 提供错误响应体的 JSON 编解码。 */
object ErrorResponse:
  given Encoder[ErrorResponse] = deriveEncoder[ErrorResponse]
  given Decoder[ErrorResponse] = deriveDecoder[ErrorResponse]
